File: common_parent.py
class TreeNode(object):
    """
    Tree node class
    """
    def __init__(self, value=None):
        """
        Constructor
        :param value:
        """
        self.left = None
        self.right = None
        self.value = value


def lowest_common_ancestor(root, node1, node2):
    result = [False, False]

    if root.value == node1.value:
        result[0] = True

    if root.value == node2.value:
        result[1] = True

    if root.left:
        left_res = lowest_common_ancestor(root.left, node1, node2)

        if isinstance(left_res, TreeNode):
            return left_res
        else:
            result = result[0] or left_res[0], result[1] or left_res[1]

    if root.right:
        right_res = lowest_common_ancestor(root.right, node1, node2)

        if isinstance(right_res, TreeNode):
            return right_res
        else:
            result = result[0] or right_res[0], result[1] or right_res[1]

    return root if result[0] and result[1] else result


if __name__ == '__main__':
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)
    n7 = TreeNode(7)
    n8 = TreeNode(8)

    n1.left = n2
    n1.right = n3
    n2.left = n4
    n2.right = n5
    n3.left = n6
    n4.left = n7
    n4.right = n8

    res = lowest_common_ancestor(n1, n6, n8)
    print(res.value)

File: common_parent.js
/*

    1
  2   3
 4 5   6
7

*/

var root = {
  value: 1,
  left: {
    value: 2,
    left: {
      value: 4,
      left: {
        value: 7
      }
    },
    right: {
      value: 5
    }
  },
  right: {
    value: 3,
    right: {
      value: 6
    }
  }
}

function findLCA(root, n1, n2) {
  let result = [false, false];
  
  if (root.value === n1) {
    result[0] = true;
  }else if (root.value === n2) {
    result[1] = true;
  }
  
  if (root.left) {
    left = findLCA(root.left, n1, n2);
    
    if (typeof(left) == 'number') {
      return left;
    } else {
      result = [result[0] || left[0], result[1] || left[1]];
    }
  }
  
  if (root.right) {
    right = findLCA(root.right, n1, n2);
    
    if (typeof(right) == 'number') {
      return right;
    } else {
      result = [result[0] || right[0], result[1] || right[1]];
    }
  }
  
  if (result[0] && result[1]) {
    return root.value;
  }
  
  return result;
}

findLCA(root, 1, 5);
